### ISSUE TITLE
remote name with dots is always interpreted as local filesystem directory

### ISSUE THREAD

**Issue by diegosainz** [rclone/rclone#3506]
#### Output of `rclone version`
```
rclone v1.48.0
- os/arch: linux/amd64
- go version: go1.12.6
```

#### Describe the issue

Given the following two remotes (in this case using backblaze backend):

```
[my.remote]
type = b2
account = xyz
key = xyz

[my-remote]
type = b2
account = xyz
key = xyz
```

When using the remote with dots in its name, it is always interpreted as a local filesystem:

```
$ rclone ls my.remote:mybucket
2019/09/03 14:01:58 ERROR : : error listing: directory not found
2019/09/03 14:01:58 Failed to ls with 2 errors: last error was: directory not found
```

Without dots it works as expected:

```
$ rclone ls my-remote:mybucket
... files....
```

BTW - Thank you for this wonderful software

**Comment by ncw** [comment]
I replicated this with the latest beta...

I'm amazed no-one has noticed this before!

**Comment by ncw** [comment]
This is because the remote names are defined like this

https://github.com/rclone/rclone/blob/b3db38ae3150aad2764a50728e9fddaa3a17ce31/fs/fspath/path.go#L14

So that is `[0-9A-Za-z_ -]` - this does not include `.`

The rclone config checks the name using the same regexp

https://github.com/rclone/rclone/blob/b3db38ae3150aad2764a50728e9fddaa3a17ce31/fs/config/config.go#L1044

```
e) Edit existing remote
n) New remote
d) Delete remote
[snip]
e/n/d/r/c/s/q> n
name> asfsdf.afsf
Can't use "asfsdf.afsf" as it has invalid characters in it.
```

So I guess the question is how did you make these remotes?

I don't think it would cause any backwards compatibility problems to add `.` to the allowed names for a remote.

**Comment by diegosainz** [comment]
Hi Nick, thank you for your quick reply and diagnose of the issue. It surprised me as well to find it, 

My use case is within a configuration management context using Ansible, where I'm using restic to configure backups for many servers and restic is using rclone as backend. Ansible is the one editing directly the `.ini` configuration file (gathering the rclone remote name, type and secrets from other sources), so it didn't go through the rclone config  process.

An alternative, fail-safe approach for this and future cases, may be:

1. If the `:` character is detected then compare the "remote" part directly with the name of configured remotes in the configuration file. 
2. If the name exists and is not a valid name (running it through the regex validation), fail with an error message. Or maybe just fail if `:` is detected and the remote name is invalid without actually checking the configuration file.

Another way, a little more expensive perhaps,  may be to simply validate all the remote names against the regex each time rclone reads its configuration and also fail in case any errors are detected. 

This way if somebody edits a remote directly in the ini file or via automated processes as in my case, it will always fail early and ensure that the remote names are valid.

**Possible backward compatibility issues**

If `.` is added to the regex then somebody that edited the remote manually or via automated processes as me, and hasn't been aware that the files are being stored locally (which is easy if the backups server has enough space and you have many remote configurations) may have a surprise to see all its files "disappear" from the remote after the upgrade (but they have been really on the local storage for the whole time).

**Comment by ncw** [comment]
> An alternative, fail-safe approach for this and future cases, may be:
> 
> 1. If the `:` character is detected then compare the "remote" part directly with the name of configured remotes in the configuration file.
> 2. If the name exists and is not a valid name (running it through the regex validation), fail with an error message. 

This is somewhat problematic because rclone needs to parse the remote:path without access to the config file (rclone can be compiled as a library without using the rclone config system).

> Or maybe just fail if `:` is detected and the remote name is invalid without actually checking the configuration file.

This might be possible... It needs a little care since rclone can deal with files with `:` - [see the docs](https://rclone.org/docs/#copying-files-or-directories-with-in-the-names).

So if the remote matched `^[^:/\\]*:` (any characters except `:`, `/`, `\` followed by a `:`) and did not match the normal remote regexp then rclone could exit with an error about an invalid remote name.

I think that still matches the docs and should stop the very suprising behaviour you saw!

> Another way, a little more expensive perhaps, may be to simply validate all the remote names against the regex each time rclone reads its configuration and also fail in case any errors are detected.

I think that is a good suggestion.  Maybe not necessary with the above though?

Rclone loads the entire config file into memory so it will be very quick to check it.

What would you think if rclone made an ERROR log but didn't exit with a fatal error?

I realise that `rclone config create` probably isn't validating the remote names either!

**Comment by diegosainz** [comment]
Oh, I didn't know about rclone being compiled as a library! - another fantastic feature :)

With that in mind I think the best approach is the one you propose:

> So if the remote matched ^[^:/\\]*: (any characters except :, /, \ followed by a :) and did not match the normal remote regexp then rclone could exit with an error about an invalid remote name.

Disregard the comparison with the config file suggestion - this way is more solid and compatible with the library approach.

> What would you think if rclone made an ERROR log but didn't exit with a fatal error?

If you meant this regarding the config file, it might be an additional tip when debugging problems (it's always good to have detailed logs on this sort of things!), but I think the approach you mention is a better solution for this issue.  I like exiting with a fatal error (and exit code > 0) because it stops many potential problems (and can be cached by monitoring systems). Since this is critical software (used for backups and all sorts of data sync) it's better to be strict regarding expected behaviour and abort on any doubts of what the user meant.

**Comment by ncw** [comment]
OK I've implemented that, plus some extra checking in `rclone config create`

I'd appreciate it if you'd give it a go and tell me what you think!

https://beta.rclone.org/branch/v1.49.0-024-g5dad88ae-fix-3506-names-beta/ (uploaded in 15-30 mins)

**Comment by diegosainz** [comment]
Nick, I'm really impressed with your attention and love to this project - only 2 days and it's solved! I really don't know how you do it in the middle of your work, family and dogs! :) Thank you.

Everything seems working as expected. I tried the following:

1. `config create`: correctly validating regex and returning error code with appropriate message in case it fails
2. Command line with invalid remote name (`ls`, `copy`, etc): correctly validating invalid remote names
3. `rclone copy ./my.dir: anotherdir`: works correctly, copying `./my.dir:` to `anotherdir`.

From my side this issue is no longer an issue :)

**Comment by ncw** [comment]
> Nick, I'm really impressed with your attention and love to this project - only 2 days and it's solved! I really don't know how you do it in the middle of your work, family and dogs! :) Thank you.

:-)

> Everything seems working as expected.

Thank you for double checking my work.  It is important to get this right!

I've just noticed the tests failed on windows so I'll revise the tests and then merge.

I'll re-open this until it is merged.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.50
